### PR TITLE
adds sentiment example for a 20b model

### PR DESCRIPTION
**Pull request by edbeeching**
This PR adds a sentiment example for a 20b model.
There are 3 scripts:
1. **s01_cm_finetune_peft_imdb.py** - Fine tuning a Low Rank Adapter on a frozen 8-bit model for text generation on the imdb dataset.
2. **s02_merge_peft_adapter.py** - Merging of the adapter layers into the base model’s weights and storing these on the hub.
3. **s03_gpt-neo-20b_sentiment_peft.py** - Sentiment fine-tuning of a Low Rank Adapter to create positive reviews.

### CODE REVIEW

File: examples/sentiment/scripts/gpt-neox-20b_peft/s03_gpt-neo-20b_sentiment_peft.py [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+
+########################################################################
+# This is a fully working simple example to use trl with accelerate.
+#
+# This example fine-tunes a GPT2 model on the IMDB dataset using PPO
+# (proximal policy optimization).
+# in any of the following settings (with the same script):
+#   - single CPU or single GPU
+#   - multi GPUS (using PyTorch distributed mode)
+#   - multi GPUS (using DeepSpeed ZeRO-Offload stages 1 & 2)
```
Comment (younesbelkada) [inline examples/sentiment/scripts/gpt-neox-20b_peft/s03_gpt-neo-20b_sentiment_peft.py:36]:
```suggestion
```

File: examples/sentiment/scripts/gpt-neox-20b_peft/s03_gpt-neo-20b_sentiment_peft.py [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+# configuration with `accelerate config`
+#
+########################################################################
+
+########################################################################
+# NOTE for to train with a 8-bit model a more recent version of
+# transformers is required, full dependecies for this example:
+# pip install  bitsandbytes datasets accelerate loralib
+# pip install  git+https://github.com/huggingface/transformers.git@main
+# pip install git+https://github.com/huggingface/peft.git
```
Comment (younesbelkada) [inline examples/sentiment/scripts/gpt-neox-20b_peft/s03_gpt-neo-20b_sentiment_peft.py:49]:
```suggestion
# pip install peft
```

File: examples/sentiment/scripts/gpt-neox-20b_peft/s03_gpt-neo-20b_sentiment_peft.py [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+def collator(data):
+    return dict((key, [d[key] for d in data]) for key in data[0])
+
+
+# set seed before initializing value head for deterministic eval
+set_seed(config.seed)
+
+# Now let's build the model, the reference model, and the tokenizer.
+# ref_model = AutoModelForCausalLMWithValueHead.from_pretrained(config.model_name, load_in_8bit=True, device_map="auto")
+# model = AutoModelForCausalLMWithValueHead.from_pretrained(config.model_name, load_in_8bit=True, device_map="auto")
```
Comment (younesbelkada) [inline examples/sentiment/scripts/gpt-neox-20b_peft/s03_gpt-neo-20b_sentiment_peft.py:140]:
```suggestion
```

File: examples/sentiment/scripts/gpt-neox-20b_peft/s03_gpt-neo-20b_sentiment_peft.py [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+
+model = AutoModelForCausalLMWithValueHead.from_pretrained(pretrained_model)
+
+model.gradient_checkpointing_disable = model.pretrained_model.gradient_checkpointing_disable
+model.gradient_checkpointing_enable = model.pretrained_model.gradient_checkpointing_enable
+
+# ref_pretrained_model = AutoModelForCausalLM.from_pretrained(config.model_name, load_in_8bit=True, device_map="auto")
+# ref_model = AutoModelForCausalLMWithValueHead.from_pretrained(ref_pretrained_model)
+
+print(model)
```
Comment (younesbelkada) [inline examples/sentiment/scripts/gpt-neox-20b_peft/s03_gpt-neo-20b_sentiment_peft.py:201]:
```suggestion
```

**Review by younesbelkada — COMMENTED** [review]
Thanks a lot for this! Left few nits! 
Let's also add a line about it on the docs 💪 Let me know if you want me to do that!

File: examples/sentiment/scripts/gpt-neox-20b_peft/README.md [hunk 5]
```diff
@@ -0,0 +1,10 @@
+
+# Fine-tuning 20B LLMs with RL on a 24GB consumer GPU
+The scripts in the section detail the fine-tuning a 20b LLM in 8-bit, in order to generate positive imdb reviews. You
+can find out more in our [blogpost](LINK_TODO).
+
+Overall there were three key steps and training scripts:
+
+1. **s01_cm_finetune_peft_imdb.py** - Fine tuning a Low Rank Adapter on a frozen 8-bit model for text generation on the imdb dataset.
+2. **s02_merge_peft_adapter.py** - Merging of the adapter layers into the base model’s weights and storing these on the hub.
+3. **s03_gpt-neo-20b_sentiment_peft.py** - Sentiment fine-tuning of a Low Rank Adapter to create positive reviews.
```
Comment (lvwerra) [inline examples/sentiment/scripts/gpt-neox-20b_peft/README.md:10]: i'd be in favour of don't introduce a numbering scheme and just keep the semantic names. small nit `cm` -> `clm` or `lm`.

I think we could try to skip the first step for large models by engineering the prompt to be something like `"A positive movie review: {PROMPT}"`. I'd guess the model would already work quite well out of the box. Maybe for a future experiment.

**Review by lvwerra — COMMENTED** [review]
In general happy to merge. Two comments
- left a nit about the naming
- we should add it to the list of examples in the docs

File: examples/sentiment/scripts/gpt-neox-20b_peft/README.md (hunk 5 above)
Comment (edbeeching) [reply examples/sentiment/scripts/gpt-neox-20b_peft/README.md]: Good point.

**Review by younesbelkada — APPROVED** [review]
Thanks a lot for this!

**Comment by Kororinpas** [comment]
Hi，there. I am a beginner in the field of NLP and I have been working with the GPT-J model recently. I came across your code for merging adapter layers into the base model's weights in s02_merge_peft_adapter.py , and I have some questions regarding the merging process.

From my understanding, after fine-tuning the model with the LORA layer and running this merging code, the LORA layer is replaced with a new randomly initialized linear layer. However, I did not see any indication in the code that the parameters of the LORA layer were inherited by this new linear layer.If this is the case, then it would mean that my previous training of only the LORA layer was pointless.

I would be grateful if you could provide me with some clarification on this matter. Thank you very much for your time and help.

**Comment by younesbelkada** [comment]
Hi @Kororinpas 
Please have a look at this nice thread: https://github.com/lvwerra/trl/issues/250 that goes over the details of merging LoRA layers! 
Also now you can directly merge the lora layers using `peft` with the `model.merge_and_unload()` utility function, see https://github.com/huggingface/peft/pull/227

**Comment by Kororinpas** [comment]
@younesbelkada Thank you for sharing this information with me. I have already checked out the thread you suggested and the original code. Now, my problem is solved. Thanks again!

**Pull request merged**